event-svc/src/event/ordering_task.rs

```diff
@@ -460,106 +539,244 @@ impl OrderingState {
+                    }
+                },
+                Err(e) => {
+                    error!(error=?e, "Failed to join task spawned to process stream batch");
+                    continue;
+                }
+            };
+            for (_, stream_events) in &streams {
                 self.deliverable
                     .extend(stream_events.new_deliverable.iter());
             }
+            processed_streams.extend(streams);
         }
 
-        match self.persist_ready_events(Arc::clone(&event_access)).await {
+        match self.persist_ready_events(&event_access).await {
             Ok(_) => {}
             Err(err) => {
-                // Clear the queue as we'll rediscover it on the next run, rather than try to double update everything.
-                // We will no-op the updates so it doesn't really hurt but it's unnecessary.
-                // The StreamEvents in our pending_by_stream map all have their state updated in memory so we can pick up where we left off.
-                self.deliverable.clear();
-                return Err(err);
+                if Self::log_error(err) {
+                    // if it's fatal we will tell the task to exit
+                    // if not, we return the state to try again
+                    return Err(());
+                } else {
+                    // Clear the queue as we'll rediscover it on the next run, rather than try to double update everything.
+                    // We will no-op the updates so it doesn't really hurt but it's unnecessary.
+                    // The StreamEvents in our pending_by_stream map all have their state updated in memory so we can pick up where we left off.
+                    self.deliverable.clear();
+                    return Ok(self);
+                }
             }
         }
         // keep things that still have missing history but don't process them again until we get something new
-        self.pending_by_stream
-            .retain(|_, stream_events| !stream_events.processing_completed());
+        for (cid, mut stream) in processed_streams.into_iter() {
+            if !stream.processing_completed() {
+                self.pending_by_stream.insert(cid, stream);
+            }
+        }
 
         debug!(remaining_streams=%self.pending_by_stream.len(), "Finished processing streams");
         trace!(stream_state=?self, "Finished processing streams");
 
-        Ok(())
+        Ok(self)
+    }
+
+    /// Splits `to_process` into chunks and spawns at most `MAX_STREAM_PROCESSING_TASKS` to review and order
+    /// the events for each chunk of streams.
+    fn spawn_tasks_for_stream_batches(
+        event_access: Arc<EventAccess>,
+        to_process: HashMap<Cid, StreamEvents>,
+    ) -> JoinSet<Result<Vec<(Cid, StreamEvents)>>> {
+        let mut task_set = JoinSet::new();
+
+        let chunk_size =
+            (to_process.len() / MAX_STREAM_PROCESSING_TASKS).max(MIN_NUM_STREAMS_PER_BATCH);
```

Comment:
to avoid spawning many tasks for small sets, we make sure they're at least 25 streams each. this choice was arbitrary and could be 50, 100, etc. with 10000 events found at once, the batches are ~600 streams each.